### ISSUE TITLE
[Declarative Beacon API] Time limit on retaining beacons for bfcache'd pages

### ISSUE THREAD

**Issue by caraitto** [WICG/unload-beacon#3]
My understanding is that with the Declarative Beacon API, if a page is kept in the bfcache for an extended period of time, it's still eligible to send the beacon once the page is evicted from the bfcache, which might be a long time after the user has navigated away from the page. 

Would it be possible to instead send the beacon for bfcache'd pages after they've been in the bfcache for a minute or so? I think the page would want to check when waking from the bfcache to see if the beacon was already sent or not (by checking the `state` field of the beacon from the explainer) -- if it's already sent, the page may want to create a new beacon (at its choice).

**Comment by caraitto** [comment]
Hi, friendly ping?

**Comment by clelland** [comment]
I think that this is what the `pageHideTimeout` setting is for -- `pageHide` is emitted when the page enters the bfcache, and that should start a timer for any beacons with `pageHideTimeout` set. If they aren't restored from cache by the time that timer expires, they will be sent at that time (rather than waiting until eviction)

**Comment by mingyc** [comment]
> I think the idea is the beacon is sent for whichever event happens first of the two?

Yes

> But, if the site doesn't set pageHideTimeout, what is the behavior?

The default behavior is specified in the `pageHideTimeout` section, i.e. If this is null when the page is hidden, the beacon will be sent on document discard (including eviction from the BFCache)

> Also, can pageHideTimeout be arbitrarily large?

It is possible as there is currently no maximum.

> maximum pageHideTimeout values to prevent sending beacons at bfcache eviction at some point long after the user has left the page.

I think users should specify pageHideTimeout if they don't want to wait too long.

**Comment by clelland** [comment]
That sounds like a second issue, @caraitto (but correct me if I'm wrong)

The first issue I see is one of developer ergonomics. In order to have beacons sent in a timely manner, there should be a way to ensure that they are sent a short time after entering the bfcache. We have that, and it's `pageHideTimeout`. My understanding is that if that is set, and a page enters the cache, then one of three things can happen: (Whichever happens first wins)
 - If the page is restored from the cache, the beacon is cancelled.
 - If the page is evicted from the cache (which browsers will do after a short timeout; the cache is fairly short-lived in current imlementations), then the beacon is sent.
 - If the page is still in the cache after `pageHideTimeout` ms, then the beacon will be sent (and will be in a 'sent' state if the page is subsequently restored)

The second issue you raise is a privacy concern, which I think is motivated by a "what if browser's *didn't* do that" question around the second point above. Currently there *is* a time limit set by browsers, which web content can't override (Chrome, for instance, has a 3 minute timer). But what if future browsers keep pages in the bfcache for an arbitrarily long time, and *also* developers don't choose to set a timeout on their own. Should we protect developers from that situation?

I feel that there could be an implementation note, certainly, that there should be *some* reasonable limit, *if* the user agent is capable of retaining cache entries indefinitely, and it warrants a section in the privacy section of the spec. I don't know if there's sufficient alignment between cache implementations to specify yet exactly what a hard limit needs to be.

**Comment by caraitto** [comment]
> I feel that there could be an implementation note, certainly, that there should be some reasonable limit, if the user agent is capable of retaining cache entries indefinitely, and it warrants a section in the privacy section of the spec. 

Yep, I think that's basically what I'd like to see -- I'll go ahead and send a PR (#5). I think 3 minutes might be OK, but I'd be less comfortable if it was over 10 minutes.

**Comment by mingyc** [comment]
This was re-discussed in https://github.com/WICG/unload-beacon/pull/16#discussion_r932857618:

- @philipwalton suggests that there are use cases longer than 10 minutes timeout
- @fergald mentions Mozilla's bfcache timeout is 45 minutes.

Maybe we should re-consider how long timeout limit should be.

**Comment by caraitto** [comment]
@fergald Continuing the discussion from https://github.com/WICG/pending-beacon/pull/16#discussion_r997429673 (seems like this is a better place to discuss):

I discussed the idea of allowing sending beacons after 10 minutes from the point a page entered the bfcache with some other Privacy / Security reviewers -- I think we're OK with this, even with frames (where the behavior introduced by beacons may not be polyfilable), because the frame site (say maps.company.com) is already able to transmit during network changes via the other tab (this is part of the general IP privacy problem).

We're also OK using eTLD+1 / First Party Sets for determining if another tab / frame is open from the same-site.

I'll update my PR -- my understanding is that if there isn't another site open with the page content, that the beacon will need to be sent within 10 minutes of entering the bfcache -- is that OK? If not, we'd need further discussion, since the privacy-wise, the API would be more like background sync (which requires a permission) -- but my understanding is that we're OK limiting to 10 minutes in the event that no other tab has a frame with the same site?

**Comment by mingyc** [comment]
1. If we lower the limit down to 10 minutes, some behavior of the API, i.e. sending on page/frame discard, will not be able to work on the browser that has longer bfcache limit (e.g. Mozilla is 45 minutes).

> if there isn't another site open with the page content, that the beacon will need to be sent within 10 minutes of entering the bfcache -- is that OK? If not, we'd need further discussion, since the privacy-wise, the API would be more like background sync (which requires a permission)
2. Current design does respect background sync permission: `Post-unload beacons are not sent if background sync is disabled for a site.`
3. (Trying to summarize) After PR https://github.com/WICG/pending-beacon/pull/41, for a beacon to live beyond the point of `pagehide` event:
   - background sync needs to be granted
   - network must be the same (currently not able to ensure this, so the Chrome implementation just sends out on pagehide).
   - it has to be < 10 minutes after pagehide
   - if > 10 minutes after pagehide, there need to be other tab/frame opening the same site.
     - => Not quite sure about this as https://github.com/WICG/pending-beacon/pull/16#discussion_r993855620 suggests `10 min limit to when no tab is active.`

Please correct if anything is wrong.

**Comment by caraitto** [comment]
> 2. Current design does respect background sync permission: `Post-unload beacons are not sent if background sync is disabled for a site.`

Oh -- this changes the privacy calculus significantly, I think, since the user *explicitly* allows the page to transmit when not active, barring that the transmission duration and frequency should be limited to avoid tracking [0].

I think we can remove the time limit requirements if the background sync permission is granted -- let me follow-up with some other reviewers offline.

...but, I'm not sure we want to *encourage* sites to ask for the background sync permission, just so they can send beacons, unless the user understands there is some clear user-focused benefit for enabling the permission (for instance, keeping content synced for offline use / startup speed boost from pre-loading). It seems to me there certainly are many sites that don't really need the background sync permission for their functionality, but they might still want to send metrics via beacons, and we want users to understand their personal risk / benefit tradeoff when granting permissions.

[0] https://wicg.github.io/background-sync/spec/#location-tracking

> 1. If we lower the limit down to 10 minutes, some behavior of the API, i.e. sending on page/frame discard, will not be able to work on the browser that has longer bfcache limit (e.g. Mozilla is 45 minutes).

Could the page go into bfcache, send the beacons 10 minutes later, and then somehow get restored after that? I think the page would need to be informed that beacons were sent while it was bfcached, in case it wants to try sending again.

> 
> > if there isn't another site open with the page content, that the beacon will need to be sent within 10 minutes of entering the bfcache -- is that OK? If not, we'd need further discussion, since the privacy-wise, the API would be more like background sync (which requires a permission)
> 
> 3. (Trying to summarize) After PR [Lower the bfcache limit to 10 min. #41](https://github.com/WICG/pending-beacon/pull/41), for a beacon to live beyond the point of `pagehide` event:
>    
>    * Background Sync permission needs to be granted. Otherwise **nothing will be sent** and all beacons get dropped.
>    * network must be the same (currently not able to ensure this, so the Chrome implementation just sends out on pagehide).
>    * it has to be < 10 minutes after pagehide
>    * if > 10 minutes after pagehide, there need to be other tab/frame opening the same site.

This SGTM (for the case where sync permission isn't granted -- I'm still following up offline on if we want to relax this when sync is granted). I assume "same site" means same eTLD+1, or in the same first party set.

**Comment by mingyc** [comment]
> ...but, I'm not sure we want to encourage sites to ask for the background sync permission, just so they can send beacons

Correct me if I am wrong, but at least in Chrome it looks like Background Sync is enabled by default for a normal user ( the setting in chrome://settings/content/backgroundSync selects the first item)

**Comment by caraitto** [comment]
My apologies, you're correct (I'm a little surprised by this) -- I discussed further offline, and everything you mentioned in https://github.com/WICG/pending-beacon/issues/3#issuecomment-1283433506 sounds correct (with exception of the question I asked about Mozilla's 45 minute time limit -- also, I'm checking if we really need to see if we're on the same network or not). So, background sync is required, and we still have the 10 min after bfcache limit, unless there's another frame of the same site (eTLD+1 / first party set).

**Comment by Sora2455** [comment]
> Background Sync permission needs to be granted. Otherwise nothing will be sent and all beacons get dropped.

Quick clarification: if a site doesn't have the background sync permission, they register a pending beacon and the user closes the page, is the beacon:

1. Sent immediately
2. Never sent at all (making the API useless)

If it's the second, how does this work in browsers that haven't implemented background sync? And could this be called out early in the explainer so that web devs will know that they can't just replace their sendBeacon + unload combinations with pending beacons, as the explainer would otherwise imply?

**Comment by mingyc** [comment]
> 1. Sent immediately
> 2. Never sent at all (making the API useless)

The current [Chrome implementation](https://chromium-review.googlesource.com/c/chromium/src/+/3874272) for Origin Trial is 2: the browser never sends any beacons if a document is about to get destroyed. 

I agree that this is not ideal. 

Looking into the wording of the explainer
> Post-unload beacons are not sent if background sync is disabled for a site.

and wording in BackgroundSync

> Recently closed sites can finish sending and receiving data

I am not sure if we can interpret the above as "allowing to send beacons before unload is finished" to allow option 1. If so, maybe we could try to send those beacons out? Need input from @caraitto as I also hope that complying to BackgroundSync could help resolve network change issue (https://github.com/WICG/pending-beacon/issues/30)

I will update the explainer to call this out as suggested in https://github.com/WICG/pending-beacon/issues/3#issuecomment-1284598728

**Comment by caraitto** [comment]
@mingyc We've discussed some more -- there are 2 separate issues:

#### A
Another live instance of the same site is present in another tab, where same site means same eTLD+1 or same first party set. Here, we probably don't need the background sync permission, since the site is still active elsewhere. However, with frames, we might have to not transmit after navigate away / close if the network changed, since with the removal of third party cookies and storage partitioning, maps.company.com in a sub-frame has a separate identity from maps.company.com in a top-level frame. This requirement to not send the beacon after page close if the network changes only needs to apply if the live instance and bfcached instance belong to separate identities. For instance, these maps.company.com instances are *each* in separate identities:

top-level: foo.com, sub-frame: maps.company.com
top-level: bar.com, sub-frame: maps.company.com
top-level: maps.company.com

But, 2 instances of maps.company.com that are (top-level: foo.com, sub-frame: maps.company.com) would be in the same identity -- likewise, 2 top level maps.company.com instances would be same identity, and therefore could send even if the network changes.

(I believe intermediate parent frames aren't part of the identity key, just the top-level and current frame).

#### B 
There is no other live instance of the same site. Transmitting beacons after navigating away from / closing the page will require the background sync permission. I'm waiting to hear from the permissions team what limits we'd like to set for this scenario.

**Comment by Sora2455** [comment]
What about a case where there is only one tab of foo.com open, but the tab navigates to another foo.com page? Does that count under scenario A or B?

**Comment by fergald** [comment]
@caraitto 
**A**
It makes perfect sense to me to apply [storage partitioning](https://github.com/privacycg/storage-partitioning) which applies to far more than just storage. I don't think we should go into details in this spec/explainer except to say that when we consider identities of sites for the purpose of decided whether it's OK to send now, we will follow storage partitioning's keying. I would also say we should relax when it is [relaxed via the API](https://github.com/privacycg/storage-partitioning).

**B**
I want to clearly distinguish 3 cases
1 sending *on* navigating away
2 sending *after* navigating away (e.g. page goes into BFCache and we wait until page destruction)
3 retrying later

IMO, 1 definitely should not require background sync. A page can already do this by adding an unload handler and sending a keepalive fetch without background sync. Sending a beacon *on* navigating away is no more powerful than this. One of the motivations for this project is to reduce unload handler usage, we really don't want to leave a case where unload+fetch is still better than beacon.

For 2, it's a bit unclear. If we enforce this then it would probably force us to send on entry to BFCache. It would definitely be better for resource usage if we could wait and see if the page is restored, knowing that we can send later if we need to. I guess we could also consider what other tabs are open, we can delay sending on BFCache as long as there is another instance open. We can live with it either way, I guess.


For 3, I think it's fine to say that this requires background sync.

@Sora2455 That should be A and I would also say that if the user goes A->B->A then any unsent beacons from the first A can be sent when second A is opened.

**Comment by caraitto** [comment]
@fergald Yep, I think we're in agreement -- the permissions team is still discussing what do do for the second of your B scenarios (2 sending after navigating away (e.g. page goes into BFCache and we wait until page destruction)).

> @Sora2455 That should be A and I would also say that if the user goes A->B->A then any unsent beacons from the first A can be sent when second A is opened.

Yep, sound good.

**Comment by mingyc** [comment]
@caraitto Is the team's discussion also covering [network changes](https://github.com/WICG/pending-beacon/issues/30) happen after navigating away?

**Comment by mingyc** [comment]
@caraitto friendly ping. Any updates from the permission team?

**Comment by fergald** [comment]
@tungnh28 So, to be explicit, in the case of no BG sync, we will still do an immediate send as we navigate away from the page. This could cause some network activity even after there is no active instance of the site. Is this in line with your expectations? The rationale here is that the page could do this anyway with a keep-alive fetch or beacon in unload/pagehide.

**Comment by mingyc** [comment]
> if we start with fully gating and observe the tendency of unload/pagehide usage, that would be better to respect user's expectation.

Just to clarify, do you suggest to start with allowing beacon sending on pagehide/unload only if BackgrounSync is enabled, even if it's achievable without using this PendingBeaconAPI?

**Comment by Sora2455** [comment]
What about browsers that don't support Background Sync and don't plan to (e.g. Firefox, Safari)? If Background Sync is required to use this API as there is no way of guaranteeing that this API won't leak network changes after page closure, does that mean that those browsers will never support Background Sync either?

**Comment by fergald** [comment]
> Not really, it's really nice if we start with fully gating and observe the tendency of unload/pagehide usage, that would be better to respect user's expectation . But if it is really the case that the gating behind permission sync could cause a tendency to increase unload handler usage as you are saying, I think it would still be ok with the current idea.

I don't expect to see unload usage increase because people are already doing keep-alive from unload/pagehide. The goal here is to provide an API that is an alternative to that but without the downsides of unload. Providing an alternative that is less capable (i.e. that will fail to send in cases where unload/pagehide would have succeeded) does not meet that goal.

If users understand unload/pagehide and keep-alive then their expectation is already that pages can send beacon-like data immediately after navigation. So doing the same with beacon is in line with user expectations.

I don't think we should hobble this API when pages can (and currently do) do this already.

**Comment by mingyc** [comment]
@caraitto 

Continuing with the scenario A&B discussed in

https://github.com/WICG/pending-beacon/issues/3#issuecomment-1286031505
https://github.com/WICG/pending-beacon/issues/3#issuecomment-1286397825

Specifically for **B-2 (sending after navigating away)** and **B-3 (retrying later)**:

- If BackgroundSync is disabled:
  - If no other live instance is open, 
    - **Should beacons all be sent before navigating away (B-1)?**
    - Or they can still be sent within **10 minutes** of entering bfcache?
  - If other live instance is open, beacons can be sent (the same as Scenario A)
- If BackgroundSync is enabled:
  -  beacons can be sent whenever

My main question is basically about whether you still want to apply 10 minutes restriction and if it's only applied to the scenario when BackgroundSync is off.

**Comment by caraitto** [comment]
@kamilahasanbega @tungnh28(permissions reviewers of this intent) to comment from the permissions / background sync side.

@mingyc Sorry I missed this -- been looking at old intents with action items for me -- not sure if anything's changed since last November? 

Paging this back in, I think we want:

> * If BackgroundSync is disabled:
>   
>   * If no other site instance is open,
>     
>     * **Should beacons all be sent before navigating away (B-1)?**

Yes

>     * Or they can still be sent within **10 minutes** of entering bfcache?

No, I think we want the background sync permission for this.

>   * If other site instance is open, beacons can be sent (the same as Scenario A)

IIUC, yes.

>     
>     * **How if other site instance was open before navigating away, but is closed after navigating away? Should beacons be dropped?**

Yes -- alternatively, they could be sent as the last other instance closes to ensure they don't get dropped.

> * If BackgroundSync is enabled:
>   
>   * beacons can be sent whenever

I don't think this is true -- I think we should only send within 10 minutes of navigating away.

**Comment by caraitto** [comment]
It looks like this was discussed here: https://github.com/WICG/pending-beacon/issues/30#issuecomment-1333869614 -- it looks like that's generally in agreement?

Originally, I thought we were thinking 40 min (this is what Firefox does) was too long, not sure if the thinking changed?

**Comment by caraitto** [comment]
Also, I think we really care about the network change requirements from #30.

**Comment by mingyc** [comment]
Update:

From the discussion in https://github.com/WICG/pending-beacon/issues/30#issuecomment-1333869614, we agree that a document needs BackgroundSync permission in order to send a pending beacon after a page is put into bfcache. If that permission is off, all pending beacons will be forced sending out right on entering bfache (e.g. navigating away).

@caraitto The 30 minutes time limit here looks irrelevant now?
